### PR TITLE
#167484555 User should be able to like and dislike articles

### PR DESCRIPTION
**Pull request by Jkadhuwa**
### What does this PR do?
- Enable authenticated users to like and dislike articles
### Description of Task to be completed?
- Create UI for Viewing single article and like and disliking articles.
- Connect the UI to the backend API
- Test the functionality

### How should this be manually tested?
- Clone this branch
- `npm install` to install all dependencies
- `npm start` to run the server
- `Signup or login to like or dislike articles`
- On the Home page, select any article and proceed to click the like or dislike buttons

### What are the relevant pivotal tracker stories?
- [Like or Dislike Articles](https://www.pivotaltracker.com/story/show/167484555)

### CODE REVIEW

**Review by harerakalex — APPROVED** [review]
LGTM

**Review by raymond42 — APPROVED** [review]
Looks fine

**Review by malaba6 — APPROVED** [review]

**Review by placideirandora — APPROVED** [review]
Looks good!

File: src/app/App.js [hunk 1]
```diff
@@ -29,13 +31,14 @@ function App() {
             component={ResetPassword}
           />
           <Route exact path="/" component={GetAllArticles} />
+          <ProtectedRoutes path="/Create" component={CreateArticle} />
```
Comment (bdushimi) [inline src/app/App.js]: @Jkadhuwa path or route names should be in lowercase

File: src/app/App.js [hunk 2]
```diff
@@ -29,13 +31,14 @@ function App() {
...
           <Route path="/login" component={Login} />
           <Route path="/signup" component={SignUp} />
           <Route path="/profile" component={Profile} />
           <Route path="/update-profile" component={UpdateProfile} />
           <Route path="/Signup" component={SignUp} />
```
Comment (bdushimi) [inline src/app/App.js]: @Jkadhuwa path or route names should be in lowercase

File: src/app/App.js [hunk 3]
```diff
@@ -29,13 +31,14 @@ function App() {
...
-          <Route path="/articles/:slug" component={SingleArticle} />
-          <ProtectedRoutes path="/Create" component={CreateArticle} />
+          {/* // <Route path="/articles/:slug" component={SingleArticle} /> */}
```
Comment (bdushimi) [inline src/app/App.js]: @Jkadhuwa You can remove this commented line if it's not used.

File: src/app/App.js (hunk 2 above)
Comment (bdushimi) [inline src/app/App.js]: @Jkadhuwa duplicated signup routes, kindly remove one of them.

File: src/feature/articles/__tests__/CreateArticleAction.test.js [hunk 4]
```diff
@@ -42,6 +42,23 @@ describe('Create Article success', () => {
...
+      response: expected
+    });
+    const mockArticles = {
+      description: 'I knew that would happen soon but i kept it inside',
+      body: 'I knew that would happen soon but i kept it inside'
+    };
+    return store.dispatch(createArticle(mockArticles)).then(() => {
+      expect(store.getActions().length).toEqual(1);
+    });
+  });
```
Comment (bdushimi) [inline src/feature/articles/__tests__/CreateArticleAction.test.js:61]: @Jkadhuwa any reason to why this PR for liking and disliking an article introduces changes in files of createArticle feature which are not related?

File: src/feature/articles/createArticle/CreateArticleComponent.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ import createArticle from './createArticleAction';
 import editorConfigs from '../../../app/config/ckeditorConfig';
 import 'react-toastify/dist/ReactToastify.css';
 import './createArticle.scss';
-import Home from '../../homePage/Home';
+// import Home from '../../homePage/Home';
```
Comment (bdushimi) [inline src/feature/articles/createArticle/CreateArticleComponent.js]: Kindly remove commented codes

File: src/feature/articles/getArticles/GetAllArticlesComponent.js [hunk 6]
```diff
@@ -2,47 +2,30 @@
...
   }
 
   render() {
-    const { loading } = this.state;
-    const { articles } = this.props;
+    const { articles, loading } = this.props;
     return (
       <>
-        <Home />
+        {/* <Home /> */}
```
Comment (bdushimi) [inline src/feature/articles/getArticles/GetAllArticlesComponent.js]: Kindly remove commented codes

File: src/feature/articles/getSingleArticle/GetSingleArticleReducer.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+export const initialState = {
+  article: {}
+};
+
+export default (state = initialState, action) => {
+  const { type, payload } = action;
+  switch (type) {
+    case LIKE_ARTICLE_SUCCESS:
+    case DISLIKE_ARTICLE_SUCCESS:
```
Comment (bdushimi) [inline src/feature/articles/getSingleArticle/GetSingleArticleReducer.js:18]: @Jkadhuwa any reason to why you're checking if some actions got dispatched and don't update the state. A reducer should always return a new state.

File: src/feature/articles/getSingleArticle/GetSingleArticleReducer.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    case GET_SINGLE_ARTICLE_SUCCESS:
+      return {
+        ...state,
+        ...payload
+      };
+    case LIKE_ARTICLE_FAIL:
+    case DISLIKE_ARTICLE_FAIL:
```
Comment (bdushimi) [inline src/feature/articles/getSingleArticle/GetSingleArticleReducer.js:25]: @Jkadhuwa any reason to why you're checking if some actions got dispatched and don't update the state. A reducer should always return a new state.

File: src/feature/articles/likeDislikeArticles/LikeDislikeAction.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+export const likeArticle = slug => async dispatch => {
+  const token = localStorage.getItem('token');
+
+  const axiosConfig = {
+    headers: {
+      'Content-Type': 'application/json',
+      Authorization: token
+    }
+  };
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/LikeDislikeAction.js]: @Jkadhuwa wouldn't be better if such axiosConfiguration is centralized i.e. put in a file and get imported wherever and whenever one needs them?

File: src/feature/articles/likeDislikeArticles/LikeDislikeAction.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+export const dislikeArticle = slug => async dispatch => {
+  const token = localStorage.getItem('token');
+
+  const axiosConfig = {
+    headers: {
+      'Content-Type': 'application/json',
+      Authorization: token
+    }
+  };
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/LikeDislikeAction.js]: @Jkadhuwa wouldn't be better if such axiosConfiguration is centralized i.e. put in a file and get imported wherever and whenever one needs them? As you can see from lines 7-15 and 34 - 41, it is duplicated.

File: src/feature/articles/likeDislikeArticles/LikeDislikeComponent.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    );
+  }
+}
+const mapStateToProps = state => ({
+  isAuthenticated: state.login.isAuthenticated
+});
+
+export default connect(
+  mapStateToProps,
+  { likeArticle, dislikeArticle }
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/LikeDislikeComponent.js]: @Jkadhuwa while passing this object which contains functions/action creators works, I suggest you use mapDispatchProps function instead.

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeAction.test.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  it('Should dispatch DISLIKE_ARTICLE_FAIL action', async () => {
+    const error = 'No articles the moment';
+    const expected = { type: DISLIKE_ARTICLE_FAIL, error };
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith(mockError(error));
+    });
+    await store.dispatch(dislikeArticle(error));
+    const dispatchedActions = store.getActions();
+    expect(dispatchedActions[1].type).toEqual(expected.type);
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeAction.test.js]: @Jkadhuwa This is test is not written correctly. You should ascertain whether store.getActions() equals expectedAction.

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeAction.test.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    const slug = 'Should dispatch GET_SINGLE_ARTICLE_SUCCESS-3244';
+
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith(mockSuccess(mockArticle.article));
+    });
+    await store.dispatch(dislikeArticle(slug));
+    const dispatchedActions = store.getActions();
+    const dispatchedTypes = dispatchedActions.map(action => action.type);
+    expect(dispatchedTypes[0]).toEqual(expected);
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeAction.test.js]: @Jkadhuwa How many actions that got dispatched when disliking an article? if there're more than one, when asserting, you should expect store.getActions().toEqual(Array of those expected Actions).

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeAction.test.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  });
+});
+describe('SERVER ERROR', () => {
+  it('Should return an error', async () => {
+    const slug = 'Should dispatch GET_SINGLE_ARTICLE_SUCCESS-3244';
+    const expected = DISLIKE_ARTICLE_FAIL;
+    await store.dispatch(dislikeArticle(slug));
+    const dispatchedActions = store.getActions();
+    const dispatchedTypes = dispatchedActions.map(action => action.type);
+    expect(dispatchedTypes[2]).toEqual(expected);
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeAction.test.js]: @Jkadhuwa How many actions that got dispatched when disliking an article? if there're more than one, when asserting, you should expect store.getActions().toEqual(Array of those expected Actions).

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    },
+    history: {
+      push: jest.fn()
+    }
+  };
+  const props = { ...initialProps, ...args };
+  return shallow(<LikeDislike {...props} />);
+};
+
+describe('Get All Articles Components tests', () => {
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js]: @Jkadhuwa the first parameter of the describe block is the name of the component under test. The component under test is LikeDislikeComponent, not GetAllArticlesComponent.

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  it('Should render a form inputs', () => {
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js]: @Jkadhuwa There're no inputs elements in LikeDislikeComponent, therefore, this test description is misleading, Kindly update.

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    likeArticle.at(0).simulate('click');
+  });
+  it('Should render a form inputs', () => {
+    const wrapper = renderLikesAndDislikes();
+    wrapper.setProps({ isAuthenticated: false });
+    const dislikeArticle = wrapper.find('.fa');
+    dislikeArticle.at(1).simulate('click');
+  });
+});
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js]: @Jkadhuwa All of the tests descriptions is this file are similar! Are they testing the same thing?

File: src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    wrapper.setProps({ isAuthenticated: false });
+    const likeArticle = wrapper.find('.fa');
+
+    likeArticle.at(0).simulate('click');
+  });
+  it('Should render a form inputs', () => {
+    const wrapper = renderLikesAndDislikes();
+    wrapper.setProps({ isAuthenticated: false });
+    const dislikeArticle = wrapper.find('.fa');
+    dislikeArticle.at(1).simulate('click');
```
Comment (bdushimi) [inline src/feature/articles/likeDislikeArticles/__tests__/LikeDislikeComponent.test.js]: @Jkadhuwa what are you expecting? The click event got simulated and nothing else! There're no assertions. It's the same for the above 3 tests, You should expect whether handleLike/handleDislike got called.

**Review by bdushimi — CHANGES_REQUESTED** [review]

File: src/feature/articles/getSingleArticle/GetSingleArticleReducer.js (hunk 7 above)
Comment (Jkadhuwa) [reply src/feature/articles/getSingleArticle/GetSingleArticleReducer.js:18]: They all do the same thing so I combined them. It works as an OR operator.

**Comment by Jkadhuwa** [comment]
@bdushimi I have finished implementing the feedback

**Review by placideirandora — CHANGES_REQUESTED** [review]
1. When I click on the like and dislike buttons while trying to view a single article, I get an error: **jwt malformed.** You should fix that.
Screen Shot 2019-10-29 at 17 17 52
Screen Shot 2019-10-29 at 17 17 45

2. I expect to start reading an article from the top (start with its title). But when click on an article (or the second one), the reading position starts in the middle and hides the article. You should fix that.
Screen Shot 2019-10-29 at 17 23 48
Screen Shot 2019-10-29 at 17 24 02

3. As you left a space between the navigation bar and the article (body), you should at the bottom as well.
Screen Shot 2019-10-29 at 17 30 13

Screen Shot 2019-10-29 at 17 30 20

4. I think the colors should be consistent with each other. Currently, Some elements are in black and others in blue.
Screen Shot 2019-10-29 at 17 30 20

5. I think the toast messages for liking/disliking and removing a like/dislike should have different background colors for the user to determine the difference easily.
Screen Shot 2019-10-29 at 17 42 40

**Comment by Jkadhuwa** [comment]
@placideirandora 
    - On `tags` that is just a placeholder as I was not working on it. 
    - On the colors, the likes are displayed as success and dislikes are displayed as a warning.

**Comment by placideirandora** [comment]
> @placideirandora
> - On `tags` that is just a placeholder as I was not working on it.
> - On the colors, the likes are displayed as success and dislikes are displayed as a warning.

Well, I have removed the tags' feedback. Try to find a way to customize the colors. If there isn't any. I will remove that feedback as well.

**Review by malaba6 — CHANGES_REQUESTED** [review]
Check the responsiveness on the mobile phone. This is what I got on a `Galaxy5` device
Screen Shot 2019-10-30 at 12 48 47

**Review by placideirandora — APPROVED** [review]

**Review by bdushimi — APPROVED** [review]

**Pull request merged**